File: src/main/scala/com/resonant/core/graph/internal/GraphConnect.java
package com.resonant.core.graph.internal;

import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;
import java.util.WeakHashMap;

/**
 * A graph that contains nodes, each with its ability to connect to other nodes.
 *
 * @author Calclavia
 */
public abstract class GraphConnect<N extends Node> implements Graph<N> {

	//A map of nodes and their connections
	protected Map<N, Set<N>> nodeMap = new HashMap<>();
	//A queue of nodes to be added to the grid upon next build()
	private Set<N> addQueue = Collections.newSetFromMap(new WeakHashMap<>());

	@Override
	public void add(N node) {
		nodeMap.remove(node);
	}

	@Override
	public void remove(N node) {
		nodeMap.remove(node);
	}

	@Override
	public Set<N> nodes() {
		return nodeMap.keySet();
	}

	@Override
	public void build() {
		addQueue.forEach(this::populate);
		addQueue.clear();
	}

	/**
	 * Populates the node list recursively
	 */
	protected void populate(N node) {
		populate(node, null);
	}

	protected void populate(N node, N prev) {
		if (!nodes().contains(node)) {
			Set<N> connections = node.connections();
			nodeMap.put(node, connections);
			connections.forEach(n -> populate(n, node));
		}
	}
}

File: src/main/scala/com/calclavia/graph/core/base/GraphConnect.scala
package com.calclavia.graph.core.base

import java.util.{List => JLIst}

import com.calclavia.graph.api.Node
import com.resonant.lib.math.matrix.AdjacencyMatrix

import scala.collection.JavaConversions._

/**
 * A graph that contains getNodes, each with its ability to connect to other getNodes.
 * @author Calclavia
 */
abstract class GraphConnect[N <: Node[_]] extends Graph[N] {
	var adjMat: AdjacencyMatrix[N] = null
	protected var nodes = Set.empty[N]

	def add(node: N) {
		nodes += node
	}

	def remove(node: N) {
		nodes -= node
	}

	//TODO: Collection?
	override def getNodes: JLIst[N] = nodes.toList

	def build() {
		adjMat = new AdjacencyMatrix[N](nodes, nodes)

		for (node <- nodes) {
			for (con <- node.connections) {
				if (nodes.contains(con)) {
					adjMat(node, con.asInstanceOf[N]) = true
				}
			}
		}
	}
}